docs/source/typing_anti_pitch.rst

```diff
@@ -0,0 +1,80 @@
+  checking usually cannot validate logic, tests can often not prove invariants of your code to
+  hold).
+
+* Your codebase is old, large and has been working fine without static type checking for years.
+  While Python's type system is designed to
+  `allow gradual adoption of static type checking <https://mypy.readthedocs.io/en/stable/existing_code.html>`_,
+  the total cost of adding type annotations to a large extant codebase can be prohibitive.
+
+* Your library does enough dynamic things that type checking would be unlikely to help your users.
+  Either you would have to spend a lot of effort redesigning your library in ways that static type
+  checkers could better understand, or you'd have to invest a lot of effort into figuring out clever
+  type annotations to twist the arms of type checkers.
+
+* Your codebase has suffered at the hands of `Hyrum's Law <https://www.hyrumslaw.com/>`_
+  and all possible observable behaviour is depended on. In order to avoid false positives for your
+  users, all your types end up being either a) complicated ``Protocol``\s that are hard to maintain,
+  or b) ``Any`` in which case there's not much point. (On the other hand, static type checking could
+  be a good solution for communicating to users what behaviour they should be allowed to rely on)
+
+* You're not opposed to type checking in theory, but you dislike Python type checkers in practice.
+  Maybe they don't understand enough of the idioms you use, maybe you'd like them to infer more
+  instead of relying on explicit annotations, maybe they're too slow, maybe they don't integrate
+  well with your editor, maybe they're too hard to configure. Whatever the reason -- it just doesn't
+  work for your project.
+
+* Type checking in Python isn't actually strict enough, powerful enough or expressive enough for
+  you. Python type checkers end up making various decisions out of pragmatism, or due to limited
+  resources, and these decisions might not be the ones for you.
+
+Advice for maintainers of untyped libraries
+*******************************************
+
+You've made the decision that adding static types isn't the right choice for your library. But
+perhaps you'd still like to help your users who do use static type checking -- and maybe you have
+some enthusiastic would-be contributors willing to help with this.
+
+One option is encourage such contributors  to publish a :pep:`561` stub-only package that is
+maintained separately from your main project. You could also contribute these stubs to the
+`typeshed <https://github.com/python/typeshed>`_ project.
+
+If more users pester you about adding static types, feel free to link them to this document. And if
+you ever change your mind, make sure to check out some of the other guides in this documentation,
+and ask any questions you have over at `Python's typing discussions <https://github.com/python/typing/discussions>`_.
```

Comment:
A lot of typing enthusiasts online have a slightly sneering attitude towards typeshed and/or the idea of stubs in general. Some of their criticisms are justified, but also, typeshed has made huge improvements in its CI checks in recent years. Possibly we could mention that there are tools available that can help to keep stubs accurate and up-to-date (maybe link to https://typing.readthedocs.io/en/latest/source/writing_stubs.html#tools-for-maintaining-stubs?)

Reply:
If the phrase "you could contribute..." means the library maintainer, then again I see no benefit over just adding the type annotations, and the point here is that the library maintainer doesn't want to do that. If it means the end user who wants there to be type annotations, then yes, but will typeshed really accept annotations for a library when the submitter isn't the library maintainer? That seems somewhat risky - once again, the example of an over-restrictive type annotation that suggests the library requires particular input types, when actually the library maintainer prefes duck typing, and fully intends (without feeling the need to explicitly say so) to allow user-defined types that conform to the interface used by the library code.

As a library author, I'd be annoyed if I found out that typeshed had published annotations that didn't match my intent, and I had to ask for their removal or improvement. The public contract of my library should be defined by me, not by someone else who adds type annotations, no matter how well-intentioned.

Reply:
https://github.com/python/typeshed/blob/main/CONTRIBUTING.md#third-party-library-stubs is the policy. I think most of stubs in typeshed are written by people separate from library maintainers.

A library can request it's stubs be changed/removed although I'm unsure that situation has happened in years typeshed existed. Stubs do sometimes leave typeshed but usually that's when library adopts the stubs and adds types.

Reply:
> If the phrase "you could contribute..." means the library maintainer, then again I see no benefit over just adding the type annotations, and the point here is that the library maintainer doesn't want to do that

I understood the phrase "You could contribute..." as referring to the external contributors pressuring the maintainer to add typing to the project. I believe that's what @hauntsaninja intended it to mean, as well. I agree that the wording is currently slightly ambiguous, though, for sure!

> will typeshed really accept annotations for a library when the submitter isn't the library maintainer?

Yes, we do this all the time, since one of the primary reasons for typeshed's existence is to ease the burden on third-party maintainers who don't want to be bothered by this kind of thing :)

> That seems somewhat risky - once again, the example of an over-restrictive type annotation that suggests the library requires particular input types, when actually the library maintainer prefes duck typing, and fully intends (without feeling the need to explicitly say so) to allow user-defined types that conform to the interface used by the library code.

Well, we review any code that's contributed to us, and any bug reports are reported to us rather than the library maintainer. Our policy is to prefer duck-type-y protocols in situations like the ones you refer to.

> As a library author, I'd be annoyed if I found out that typeshed had published annotations that didn't match my intent, and I had to ask for their removal or improvement. The public contract of my library should be defined by me, not by someone else who adds type annotations, no matter how well-intentioned.

I'm sorry to hear that. We're generally pretty receptive to bug reports, whether they come from users or maintainers of libraries. We also generally remove stubs if a package has added a `py.typed` file to their package, indicating that they now either have inline types or ship their own stubs, rendering typeshed's stubs obsolete. I can't recall a situation where a maintainer has asked for us to remove stubs from typeshed without them adding a `py.typed` file first. If that happened, it would probably be tricky for us to balance the needs of our users (who might now be depending on a stubs package we're publishing to PyPI), and the desire of the maintainer.

Note, however, that even if typeshed didn't exist, it's still within anybody's power to publish a stubs package for any project they like! I think that typeshed is pretty clear about the fact that our stubs packages are contributed by third-party users, shouldn't be seen as "official" (except perhaps for the stdlib), and aren't necessarily blessed by the authors of the runtime package. Our README states:

> Typeshed contains external type annotations for the Python standard library and Python builtins, as well as third party packages as contributed by people external to those projects.

So, I think it's perhaps a stretch to say that typeshed claims to attempt to define the "public contract" of any external library. We're just doing the best we can to help people out in situations where users want typing but maintainers don't want to add it :)

Reply:
> So, I think it's perhaps a stretch to say that typeshed claims to attempt to define the "public contract" of any external library.

That's fair. I'm very much talking theoretically at this point. At this point, the few libraries I publish personally have type annotations, and for the ones I contribute to adding types isn't my sole decision (and they mostly have adopted type annotations too).

The problem is that a lot of this is (as I've mentioned elsewhere, and you alluded to above) very much a social problem. It's about people's attitudes, and assumed responsibilities. And one of the attitudes that I'm bringing into this mix is a certain level of defensiveness about "ownership" of projects that I consider "mine". This comes from a history of me now having very limited volunteer time, and encountering too many users who feel they have a right to dictate my priorities. As a result, I'm taking a lot of this discussion too personally. Sorry for that.

In practice, all I really want here is for the document to be clearer that the suggestion is that *the user* can offer their suggested stubs to typeshed. That could be as simple as saying "*They* could also contribute...".

Reply:
> The problem is that a lot of this is (as I've mentioned elsewhere, and you alluded to above) very much a social problem. It's about people's attitudes, and assumed responsibilities. And one of the attitudes that I'm bringing into this mix is a certain level of defensiveness about "ownership" of projects that I consider "mine". This comes from a history of me now having very limited volunteer time, and encountering too many users who feel they have a right to dictate my priorities. As a result, I'm taking a lot of this discussion too personally. Sorry for that.

Absolutely, I completely get that. As a typeshed maintainer, I view typeshed as part of the solution here, since one of the things we're trying hard to do is to reduce the burden on library maintainers like yourself. But we could also probably be even clearer about the fact that our third-party stubs are generally "unsactioned" :)

Reply:
@hauntsaninja, you marked this thread as "resolved", but none of the discussion on this thread has actually touched on my original suggestion in https://github.com/python/typing/pull/1477#discussion_r1335520420 that we could add a link here to the docs for best practices on testing stubs. The discussion has all been addressing Paul's comment in reply, which was really about a pretty different issue :)

(Though maybe you also think my suggestion of linking to best practices for testing stubs is a bad idea ;)

Reply:
Sorry, missed that amongst the long replies in this thread, linked https://github.com/python/typing/pull/1477/commits/2836871dd559f0d7034407be36a9ba5fd2e0417d